src/Npgsql/NpgsqlConnector.cs

```diff
@@ -249,6 +254,30 @@ internal bool PostgresCancellationPerformed
 
         internal readonly Stopwatch QueryLogStopWatch = new();
 
+        static readonly string[] CriticalFailureCodes =
+        {
+            PostgresErrorCodes.ConnectionException,
+            PostgresErrorCodes.ConnectionDoesNotExist,
+            PostgresErrorCodes.ConnectionFailure,
+            PostgresErrorCodes.ProtocolViolation,
+            PostgresErrorCodes.DiskFull, // Self explanatory
+            PostgresErrorCodes.OutOfMemory, // Self explanatory
+            PostgresErrorCodes.TooManyConnections,
```

Comment:
Note that unlike the rest of the errors here, TooManyConnections shouldn't cause any issues with already existing pooled connections (the other errors probably mean that all other connections are dead too). This probably means that all errors *except* this one should clear the idle connections in the pool (#3441), but TooManyconnections definitely should not clear the pool.

Reply:
Agree, but clearing the pool should be done in another pr.

Reply:
Agreed. Made a note: https://github.com/npgsql/npgsql/pull/3441#issuecomment-805919356.